### PR TITLE
CMOR check generic level coordinates in CMIP6

### LINKED ISSUE #516: CMOR checker does not check CMIP6 generic levels
The CMOR check does not check generic levels, see e.g.
https://github.com/ESMValGroup/ESMValCore/blob/c460ceafde91014df67a67d7f7de8a0867a339ac/esmvalcore/cmor/check.py#L333

However, generic level definitions are available in the CMIP6_coordinates.json file, e.g. (notice line 500)
https://github.com/ESMValGroup/ESMValCore/blob/c460ceafde91014df67a67d7f7de8a0867a339ac/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_coordinate.json#L479-L501

Therefore it would be good if the checker could be improved so it checks these coordinates if a definition is available.

### PR DESCRIPTION
**Pull request by sloosvel**
For each generic level, all coordinates that have the same generic_level_name get loaded when reading the tables under a new CoordinateInfo attribute. The checker checks if the out_name and standard_name match any of the possible coordinates, and if so proceeds with the checks with the right coordinate. 

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #516

### CODE REVIEW

**Comment by rswamina** [comment]
@sloosvel  - Am I correct in understanding that these fixes will fix the problem of variables having different aliases in standard names (e.g. mrsos has moisture_content_of_soil_layer
or mass_content_of_water_in_soil_layer) as well as the problem that some models like IPSL have where ocean levels is not "plev" or "lev" but "olev"? I will be interested in knowing when this branch will be merged.

**Comment by sloosvel** [comment]
>  fix the problem of variables having different aliases in standard names (e.g. mrsos has moisture_content_of_soil_layer or mass_content_of_water_in_soil_layer) 

Not really. This pull request does not affect or modify the checking of variable's standard names, it just adds a check for generic levels in CMIP6. 

 > the problem that some models like IPSL have where ocean levels is not "plev" or "lev" but "olev"

Thanks for the reference, I checked the dataset and there are so many wrong things with it. It seems like it's not reporting the error because I assumed that the coordinate would have at least a standard_name, and it that case it does not even have that. I will need to rework the code a bit to make it work.

**Comment by rswamina** [comment]
@jvegasbsc and @sloosvel - I followed the thread of an original ESMValTool issue regarding variable alias management to this one, via issue https://github.com/ESMValGroup/ESMValTool/pull/950 . This is with reference to the soil moisture variable mrsos having two aliases for short name - "moisture_content_of_soil_layer" and "mass_content_of_water_in_soil_layer". The UKESM and MRI model data for mrsos are two examples where the latter alias causes the esmvaltool preprocessor to fail. Will these be fixed ?

**Comment by sloosvel** [comment]
I think it's a mistake on the reference, this pull request has nothing to do with the issue with the aliases. Maybe you are looking for #595 ?

File: esmvalcore/cmor/check.py [hunk 1]
```diff
@@ -380,11 +381,52 @@ def _check_dim_names(self):
                             self.report_error(
                                 self._does_msg, coordinate.name, 'exist')
 
+    def _check_generic_level_dim_names(self, key, coordinate):
+        if coordinate.generic_lev_coords:
+            for (_, coord) in coordinate.generic_lev_coords.items():
```
Comment (bouweandela) [inline esmvalcore/cmor/check.py:386]:
```suggestion
            for coord in coordinate.generic_lev_coords.values():
```

**Comment by bouweandela** [comment]
@jvegasbsc @mattiarighi Today is the deadline for anything that is included in the first version 2 release. Would you have time to review/test this today? If not, this can go into the next release.

**Comment by bouweandela** [comment]
I just tried to test this and got the following error message:
```
thetao: does not match coordinate rank
```
Is there indeed an error in this dataset, or is the check too strict?
[ncdump of the input data](https://github.com/ESMValGroup/ESMValCore/files/4859073/ncdump.txt)
[recipe_python.yml](https://github.com/ESMValGroup/ESMValCore/files/4859075/recipe_python.yml.txt)

**Comment by sloosvel** [comment]
I guess this is because of this attribute?

´thetao:coordinates = "lat lon" ;´ . It's missing the depth.

**Comment by jvegasbsc** [comment]
> 
> 
> I guess this is because of this attribute?
> 
> `thetao:coordinates = "lat lon" ;`. It's missing the depth.

It is ok, lev is a dimension, so you don't need to specify it in the coordinates

**Comment by jvegasbsc** [comment]
@sloosvel , take a look at `check_rank`:

https://github.com/ESMValGroup/ESMValCore/blob/94568f89e80aa0da995265dc9518f4206b38ad10/esmvalcore/cmor/check.py#L297

Maybe `var_info.coords` contains all possible generic levels, making it expect too much dimensions?

**Comment by jvegasbsc** [comment]
Ok, I see it is exacatly the opposite. Is not checking the generic levels so the expected rank is one dimension less

**Comment by sloosvel** [comment]
It's the last commit with the changes in the for loop when reading the table. For some reason it's assigning to the olevel coordinate info a time dimension information. Therefore is not being recognised as a generic_level coord. Reversing the commit the error that appears for the dataset is the following:

`esmvalcore.cmor.check.CMORCheckError: There were errors in variable thetao:
Coordinate olevel has wrong standard_name.`

**Comment by jvegasbsc** [comment]
> It's the last commit with the changes in the for loop when reading the table. For some reason it's assigning to the olevel coordinate info a time dimension information. Therefore is not being recognised as a generic_level coord. Reversing the commit the error that appears for the dataset is the following:

So revert the commit, the standard_name is empty in the dataset, so it should be an error. Maybe we can improve the error message

**Comment by sloosvel** [comment]
Do I have to adress anything else here? I have a codacy error but I think it can't be easily solved.

**Comment by bouweandela** [comment]
We would need someone to test this with some real CMIP6 data, ideally running one or more recipes that use a lot of CMIP6 data to see if it works as expected. @ESMValGroup/esmvaltool-developmentteam Any volunteers?

**Comment by schlunma** [comment]
> We would need someone to test this with some real CMIP6 data, ideally running one or more recipes that use a lot of CMIP6 data to see if it works as expected. @ESMValGroup/esmvaltool-developmentteam Any volunteers?

I can do that

**Comment by schlunma** [comment]
I get error messages like this now

```
esmvalcore.cmor.check.CMORCheckError: There were errors in variable cl:
lev: has values > valid_max = 1.0
```

which do not appear when I use the current master branch. Unfortunately I cannot trace them to changes in this PR, any ideas?

Note that I merged the current master branch into this one (locally) before testing.

**Comment by sloosvel** [comment]
Could you let me know which dataset is that?

**Comment by schlunma** [comment]
I had the issue with CAS-ESM2-0 (`CAS/CAS-ESM2-0/historical/r1i1p1f1/Amon/cl/gn/v20200302/cl_Amon_CAS-ESM2-0_historical_r1i1p1f1_gn_185001-201412.nc`) and UKESM1-0-LL (`MOHC/UKESM1-0-LL/historical/r1i1p1f2/Amon/cl/gn/v20190406/cl_Amon_UKESM1-0-LL_historical_r1i1p1f2_gn_195001-199912.nc`), but I guess more are affected.

The variable `cl` uses a hybrid height/pressure coordinate (see e.g. https://scitools.org.uk/iris/docs/latest/iris/iris/aux_factory.html), so the actual value of `lev` does not really matter.

**Comment by sloosvel** [comment]
I'm not able to reproduce the error for ´cl´. The generic_level check starts checking against an undetermined generic level coordinate:

```
>>>self._cmor_var.coordinates.keys()

dict_keys(['longitude', 'latitude', 'alevel', 'time'])```

```
By checking the cube against all possible `alevel` coordinates , since the cube metadata is already correct, it's able to guess that the  coordinate is `hybrid_height`:

```
>>>self._cmor_var.coordinates.keys()

dict_keys(['longitude', 'latitude', 'time', 'hybrid_height'])
```
Which makes sense for `cl`. So the checker continues knowing that the level coordinate from the cube has to be checked against the `hybrid_height` information, including that there is not a `valid_max` value:

```
>>>self._cmor_var.coordinates['hybrid_height'].valid_max

''
```

And adding an additional debug message:

```
2020-10-05 07:11:59,132 UTC [13887] DEBUG   There were metadata changes in variable cl:
Generic level coordinate alevel will be checked against hybrid_height coordinate information.

```

So the checker gives no further errors, at least for me.  Even when merging the latest changes in master.

**Comment by schlunma** [comment]
That is weird. Can you merge the current `master` branch into this one and test again? I did this locally on my machine before testing. Note that when I'm using the current `master` branch I do **not** get this error.

**Comment by schlunma** [comment]
Sorry, ignore my previous message, I just saw that you already merged `master` into it but I didn't see the commit message here.

I will test again.

**Comment by schlunma** [comment]
Still the same error. Here are the recipe and the log file:

[recipe.txt](https://github.com/ESMValGroup/ESMValCore/files/5326112/recipe.txt)
[out.txt](https://github.com/ESMValGroup/ESMValCore/files/5326109/out.txt)

**Comment by jvegasbsc** [comment]
I have been able to reproduce your error on Jasmin. I am downloading the dataset to debug it locally

**Comment by schlunma** [comment]
Can you also confirm that the error does not appear in the current `master`?

**Comment by jvegasbsc** [comment]
> Can you also confirm that the error does not appear in the current master?

Yes, no error in master

**Comment by jvegasbsc** [comment]
This is weird, I just realized that UKESM1-0-LL  is working for me, but HadGEM3-GC31-LL is failing

**Comment by schlunma** [comment]
> This is weird, I just realized that UKESM1-0-LL is working for me, but HadGEM3-GC31-LL is failing.

Apparently I mixed up the UK models, UKESM1-0-LL is also working for me and HadGEM3-GC31-LL is not. Sorry for the confusion!

**Comment by jvegasbsc** [comment]
> Apparently I mixed up the UK models, UKESM1-0-LL is also working for me and HadGEM3-GC31-LL is not. Sorry for the confusion!

Can you please test them isolated in the recipe? Both work for me when run alone...

**Comment by sloosvel** [comment]
Both UKESM1-0-LL and  HadGEM3-GC31-LL work for me. CAS-ESM which is the one that is failing in that log is not on Jasmin. But looking at the log, that dataset has the standard name for the level set to `atmosphere_hybrid_sigma_pressure_coordinate` instead of `"atmosphere_hybrid_height_coordinate"`, as in UKESM and HadGem.
And the information for `atmosphere_hybrid_sigma_pressure_coordinate` in the tables does have a valid_max value set to 1.0. So the checker is right to fail.

**Comment by valeriupredoi** [comment]
> > This is weird, I just realized that UKESM1-0-LL is working for me, but HadGEM3-GC31-LL is failing.
> 
> Apparently I mixed up the UK models, UKESM1-0-LL is also working for me and HadGEM3-GC31-LL is not. Sorry for the confusion!

How dare you? :gb: UKESM means UKESM :grin: 
You guys need help with this?

**Comment by schlunma** [comment]
> > Apparently I mixed up the UK models, UKESM1-0-LL is also working for me and HadGEM3-GC31-LL is not. Sorry for the confusion!
> 
> Can you please test them isolated in the recipe? Both work for me when run alone...

They both work for me too...weird. I was pretty sure that one of the UK models was failing. Unfortunately I don't have the log file anymore.



> Both UKESM1-0-LL and HadGEM3-GC31-LL work for me. CAS-ESM which is the one that is failing in that log is not on Jasmin. But looking at the log, that dataset has the standard name for the level set to `atmosphere_hybrid_sigma_pressure_coordinate` instead of `"atmosphere_hybrid_height_coordinate"`, as in UKESM and HadGem.
> And the information for `atmosphere_hybrid_sigma_pressure_coordinate` in the tables does have a valid_max value set to 1.0. So the checker is right to fail.

I agree, the checker is probably right to fail, but why does it do that in this branch and not in master? I do not see any changes in this branch regarding the check for `valid_min` and `valid_max`.

**Comment by sloosvel** [comment]
Because in master the level coordinate is not being checked because of:

```
           # Cannot check generic_level coords as no CMOR information
            if coordinate.generic_level:
                continue
```

The point of this pull request is to start checking generic level coordinates. Which probably means more errors will be reported, considering how "diverse" datasets are.

**Comment by jvegasbsc** [comment]
I am bit worried because `esmvaltool run ~/recipe_generic_levels.yml --max-years 2 --skip-nonexistent` with the complete example that you send HadGEM fails. If I comment all the others, it goes thorugh, both with --max-years and without

**Comment by schlunma** [comment]
> I am bit worried because `esmvaltool run ~/recipe_generic_levels.yml --max-years 2 --skip-nonexistent` with the complete example that you send HadGEM fails. If I comment all the others, it goes thorugh, both with --max-years and without

Can confirm. Setting the years to 1980-1980 though will also trigger this error when all datasets are used. UKESM shows the same behavior

**Comment by schlunma** [comment]
> Because in master the level coordinate is not being checked because of:
> 
> ```
>            # Cannot check generic_level coords as no CMOR information
>             if coordinate.generic_level:
>                 continue
> ```
> 
> The point of this pull request is to start checking generic level coordinates. Which probably means more errors will be reported, considering how "diverse" datasets are.

Ahh, alright, that makes totally sense. Sorry for the misunderstanding. So I guess this PR is working.

However, I'm not sure if we should merge this now without providing fixes to the corresponding datasets, since this PR will introduce new errors that were not present in the old `master`.  However, I do not know how to fix these issues (i.e. how to reasonably project the coordinate values outside of [0,1] into [0,1]).

**Comment by schlunma** [comment]
Found 3 more errors (for the "full" time range 1980-2014 but also for 1980-1980):

CESM2

```
esmvalcore.cmor.check.CMORCheckError: There were errors in variable cl:
lev: units should be 1, not hPa
 lev: has values < valid_min = 0.0
```

CESM2-WACCM

```
esmvalcore.cmor.check.CMORCheckError: There were errors in variable cl:
lev: units should be m, not hPa
 lev: has values < valid_min = 0.0
```

CESM2-FV2

```
esmvalcore.cmor.check.CMORCheckError: There were errors in variable cl:
lev: units should be 1, not hPa
 lev: has values < valid_min = 0.0
```

**Comment by bouweandela** [comment]
> However, I'm not sure if we should merge this now without providing fixes to the corresponding datasets

Good point. Let's wait with merging this until the feature freeze (@valeriupredoi and I will do that around 3 PM CEST today), if everything is working as expected here we can get this merged after and have an opportunity to implement the required fixes in the next few months, before the next release.

**Comment by sloosvel** [comment]
>I am bit worried because esmvaltool run ~/recipe_generic_levels.yml --max-years 2 --skip-nonexistent with the complete example that you send HadGEM fails. If I comment all the others, it goes thorugh, both with --max-years and without

I think there is an ambiguity problem with this particular variable. HadGem or UKESM have the `alevel` coordinate set to `atmosphere_hybrid_height_coordinate`, whereas dataset BCC-ESM1 has it set to `atmosphere_hybrid_sigma_pressure_coordinate`.
And those two coordinates have different standards regarding units, valid_min and max, etc. There is no way to know beforehand if the `alevel` is one or the other, because both are valid standard names for `alevel`. 

If the recipe only contains datasets that share the standard name, it will pass the checks without problems.   HadGem and UKESM in the coordinate for `cl` will be assumed to be `atmosphere_hybrid_height_coordinate`, and since that is the case for both datasets, the recipe will finish succesfully.

If you run a recipe with BCC and HadGem (different standard names), the checker will assume that the right coordinate is the one from the dataset that checks first.  So if BCC is checked first, HadGem will fail because the alevel coordinate will checked against to `atmosphere_hybrid_sigma_pressure_coordinate`. 

So what should be done in those cases? Overwrite the cmor information for every dataset?

**Comment by jvegasbsc** [comment]
> So what should be done in those cases? Overwrite the cmor information for every dataset?

Yes, the check should be independent for each dataset.

**Comment by sloosvel** [comment]
With the latest changes,  I can run UKESM and HadGEM without failing anymore when called with other models that have a different standard name for the lev variable . 

CESM2 is not passing the checker, but that is expected with the new checks because there is not any `alevel` coordinate with units of 'hPa' and the valid_min is '0.0' for all `alevel` . Other CESM datasets I do not have access to, but I would expect them to make the checker fail because of the same reason as CESM2.  So fixes for those datasets should be added.

I also don't have access to CAMS-CSM1-0, CAS-ESM2-0, CMCC-CM2-SR5, E3SM-1-0,  FGOALS-g3, GISS-E2-1-H, MIROC-ES2L,  MPI-ESM-1-2-HAM, MPI-ESM1-2-LR, NESM3, NorESM2-LM, NorESM2-MM, SAM0-UNICON and TaiESM1. But I will try to test them tomorrow.

**Comment by valeriupredoi** [comment]
picking up from #859 for CESM2/thetao I am getting this:
```
esmvalcore.cmor.check.CMORCheckError: There were errors in variable thetao:
Coordinate olevel has wrong standard_name or is not set.
in cube:
sea_water_potential_temperature / (degC) (time: 1980; lev: 60; cell index along second dimension: 384; cell index along first dimension: 320)
```
loading the input data off ESGF:
```python
c = iris.load_cube("/badc/cmip6/data/CMIP6/CMIP/NCAR/CESM2/historical/r1i1p1f1/Omon/thetao/gn/latest/thetao_Omon_CESM2_historical_r1i1p1f1_gn_185001-201412.nc")
```
indeed for `lev` thse standard_name is None - good work @sloosvel - can we merge this now or are we waiting for anything else? Note that I ran it through various check levels and they behave as expected :+1: Also ran `recipe_python.yml` and another recipe with UKESM1-0-LL - no problemos! :beer:

**Review by valeriupredoi — APPROVED** [review]
let's merge this, it's a very useful feature :+1:

**Comment by sloosvel** [comment]
The idea was to add fixes before merging I think. Because some of the coordinates like in #840 have wrong values and it has gone undetected. But in some cases it's not that straightforward to fix.

**Comment by valeriupredoi** [comment]
no no - it is beyond the scope of this PR to add fixes, plus, it'd bloat it out of proportion - this is a checker enhancement, if you can add an automated fix in `_fix.py` then yeah but it's not that straightforward; e.g. I will add a fix for CESM2 as per #859 but only after this gets merged

**Comment by sloosvel** [comment]
Then I guess it can be merged? I think that the problem with mixing datasets above got solved as well.

**Comment by valeriupredoi** [comment]
yep - @bouweandela and @jvegasbsc as reviewers you OK with merge? @schlunma as well, since you've noticed that issue?

**Comment by jvegasbsc** [comment]
Yes, go ahead

**Comment by valeriupredoi** [comment]
done, cheers Saskia and Javi! Manu man, if you still see that issue, forget about it - just kidding, open an Issue pls, bro :beer:

**Comment by schlunma** [comment]
I think we should have waited with the merge...That PR broke at least one recipe (https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/recipes/recipe_ecs_constraints.yml) and maybe others that we are not aware of.

**Comment by valeriupredoi** [comment]
@schlunma this is a chicken-egg situation - we need to fry the chicken now ie fix the data that gets into those recipes because inherently those recipes may be wrong since the data that gets in and analyzed may be wrong. Cheers for opening the issue! :beer: But an enhanced functionality in Core should not wait on the fix(es) needed in Tool :+1:

**Comment by schlunma** [comment]
I still have this weird issue that (I think) depends on the order of which the datasets are read:

The following recipe sometimes works, sometimes fails with 

```
esmvalcore.cmor.check.CMORCheckError: There were errors in variable cl:
lev: units should be 1, not m
 lev: has values > valid_max = 1.0
```

for `HadGEM3-GC31-MM` and sometimes with the same error for `ACCESS-ESM1-5`.

Each of theses dataset does **NOT** throw an error when used alone.

What is going on here???

The recipe:

```yaml
# ESMValTool
---
documentation:
  authors:
    - schlund_manuel

  maintainer:
    - schlund_manuel

  references:
    - gregory04grl

  projects:
    - crescendo


diagnostics:

  diag_test:
    variables:
      cl:
        project: CMIP6
        exp: historical
        mip: Amon
        start_year: 2004
        end_year: 2004
    additional_datasets:
      - {dataset: ACCESS-ESM1-5,   ensemble: r1i1p1f1, grid: gn}
      - {dataset: GISS-E2-1-G,     ensemble: r1i1p1f1, grid: gn}
      - {dataset: GISS-E2-1-H,     ensemble: r1i1p1f1, grid: gn}
      - {dataset: HadGEM3-GC31-LL, ensemble: r1i1p1f3, grid: gn}
      - {dataset: HadGEM3-GC31-MM, ensemble: r1i1p1f3, grid: gn}
    scripts:
      null
```

**Comment by bouweandela** [comment]
@schlunma To create more visibility for this, it might be good to create an issue for that instead of commenting on closed pull request.

**Comment by schlunma** [comment]
Makes sense, see #883

**Pull request merged**